cli/command/secret/utils.go

```diff
@@ -1,47 +1,81 @@
 package secret
 
 import (
+	"strings"
+
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
 	"github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/client"
 	"golang.org/x/net/context"
 )
 
-func getSecretsByName(ctx context.Context, client client.APIClient, names []string) ([]swarm.Secret, error) {
+func getSecretsByNameOrIDPrefixes(ctx context.Context, client client.APIClient, terms []string) ([]swarm.Secret, error) {
 	args := filters.NewArgs()
-	for _, n := range names {
+	for _, n := range terms {
 		args.Add("names", n)
+		args.Add("id", n)
 	}
 
 	return client.SecretList(ctx, types.SecretListOptions{
 		Filters: args,
 	})
 }
 
-func getCliRequestedSecretIDs(ctx context.Context, client client.APIClient, names []string) ([]string, error) {
-	ids := names
-
-	// attempt to lookup secret by name
-	secrets, err := getSecretsByName(ctx, client, ids)
+func getCliRequestedSecretIDs(ctx context.Context, client client.APIClient, terms []string) ([]string, error) {
+	secrets, err := getSecretsByNameOrIDPrefixes(ctx, client, terms)
 	if err != nil {
 		return nil, err
 	}
 
-	lookup := make(map[string]struct{})
-	for _, id := range ids {
-		lookup[id] = struct{}{}
+	lookup := make(map[string]bool)
+	for _, term := range terms {
+		lookup[term] = false
 	}
 
 	if len(secrets) > 0 {
-		ids = []string{}
+		found := make(map[string]struct{})
 
 		for _, s := range secrets {
-			if _, ok := lookup[s.Spec.Annotations.Name]; ok {
-				ids = append(ids, s.ID)
+			// attempt to lookup secret by full ID
+			if v, ok := lookup[s.ID]; ok && !v {
+				found[s.ID] = struct{}{}
+				lookup[s.ID] = true
 			}
 		}
+
+		for _, s := range secrets {
+			// attempt to lookup secret by full name
+			if v, ok := lookup[s.Spec.Annotations.Name]; ok && !v {
+				found[s.ID] = struct{}{}
+				lookup[s.Spec.Annotations.Name] = true
+			}
+
+		}
+
+		// attempt to lookup secret by partial ID (prefix)
+		for term, v := range lookup {
+			// Only perform the check if both full ID and full name above return no match
+			if !v {
+				for _, s := range secrets {
+					if strings.HasPrefix(s.ID, term) {
+						found[s.ID] = struct{}{}
+						lookup[term] = true
```

Comment:
why do we need to set this to `true` if there is no other loop/interaction with `lookup` later on?